### PR TITLE
Name compliance check

### PR DESCRIPTION
**Pull request by ober37**

### CODE REVIEW

File: exporter.go [hunk 1]
```diff
@@ -24,6 +24,8 @@ const (
 	SPConnectionCount      string = "sp_connection_established_count"
 	SPDisconnectionCount   string = "sp_connection_lost_count"
 
+	SPPushNewMetric        string = "sp_new_metrics_pushed"     //# adding new unknown metric
```
Comment (ober37) [inline exporter.go]:
```suggestion
	SPPushNewMetric        string = "sp_invalid_metric_name_received"     //# adding new unknown metric
```

Comment (ober37) [reply exporter.go]: Also, remove the "#" from the comment

File: exporter.go [hunk 2]
```diff
@@ -198,11 +200,25 @@ func (e *spplugExporter) receiveMessage() func(mqtt.Client, mqtt.Message) {
...
 		// Sparkplug messages contain multiple metrics within them
 		// traverse them and process them
+
 		metricList := pbMsg.GetMetrics()
 
 		for _, metric := range metricList {
-			metricName := metric.GetName()
 
+			metricName, err := getMetricName(metric)
+			if  err != nil{
```
Comment (ober37) [inline exporter.go]:
```suggestion
			if  err != nil {
```

File: exporter.go [hunk 3]
```diff
@@ -198,11 +200,25 @@ func (e *spplugExporter) receiveMessage() func(mqtt.Client, mqtt.Message) {
...
+				log.Debugf("Error: %s %v  \n", metricName, err)
```
Comment (ober37) [inline exporter.go]:
```suggestion
				log.Errorf("Error: %s %v  \n", metricName, err)
```

Comment (ober37) [inline exporter.go]:
```suggestion
				log.Errorf("Error: %s %s %v  \n", labelValues["sp_edge_node_id"], metricName, err)
```

Comment (ober37) [reply exporter.go]: Increment the new "bad metric" stat here in this block

File: exporter.go [hunk 4]
```diff
@@ -198,11 +200,25 @@ func (e *spplugExporter) receiveMessage() func(mqtt.Client, mqtt.Message) {
...
+			}else{
```
Comment (ober37) [inline exporter.go]:
```suggestion
			} else {
```

File: exporter.go [hunk 5]
```diff
@@ -198,11 +200,25 @@ func (e *spplugExporter) receiveMessage() func(mqtt.Client, mqtt.Message) {
...
+				eventString = "Unknown metric"
```
Comment (ober37) [inline exporter.go]: Not needed

File: exporter.go [hunk 6]
```diff
@@ -198,11 +200,25 @@ func (e *spplugExporter) receiveMessage() func(mqtt.Client, mqtt.Message) {
...
+				if metricVal, err := convertMetricToFloat(metric); err == nil{
```
Comment (ober37) [inline exporter.go]: Remove this entire block

File: exporter.go [hunk 7]
```diff
@@ -198,11 +200,25 @@ func (e *spplugExporter) receiveMessage() func(mqtt.Client, mqtt.Message) {
...
+					log.Debugf("%s %s : %g\n", eventString, metricName, metricVal)
+					e.metrics[metricName].With(labelValues).Set(metricVal)
+					e.metrics[SPLastTimePushedMetric].With(labelValues).SetToCurrentTime()
+					e.counterMetrics[SPPushNewMetric].With(labelValues).Inc()
+				}
+			}
+
+			// fmt.Println(matchRes)
```
Comment (ober37) [inline exporter.go]: remove commented out code

File: utilities.go [hunk 8]
```diff
@@ -3,6 +3,8 @@ package main
 import (
 	"errors"
 	"strings"
+	// "fmt"
```
Comment (ober37) [inline utilities.go]: remove this.

File: utilities.go [hunk 9]
```diff
@@ -111,6 +113,21 @@ func getNodeLabelSet() []string {
 	return []string{SPNamespace, SPGroupID, SPEdgeNodeID}
 }
 
+func getMetricName(metric *pb.Payload_Metric) (string, error){
```
Comment (ober37) [inline utilities.go]:
```suggestion
func getMetricName(metric *pb.Payload_Metric) (string, error) {
```

File: utilities.go [hunk 10]
```diff
@@ -111,6 +113,21 @@ func getNodeLabelSet() []string {
...
+
+	metricName := metric.GetName()
+	var errUnexpectedType = errors.New("Metrics does not matches with prometheus")
+	match, _ := regexp.MatchString("[a-zA-Z_:][a-zA-Z0-9_:]*", metricName)
+	if match == true{
+		return metricName, nil
```
Comment (ober37) [inline utilities.go]: Remove the debug print, let's the caller report this.

File: utilities.go [hunk 11]
```diff
@@ -111,6 +113,21 @@ func getNodeLabelSet() []string {
 	return []string{SPNamespace, SPGroupID, SPEdgeNodeID}
 }
 
+func getMetricName(metric *pb.Payload_Metric) (string, error){
+
+	metricName := metric.GetName()
+	var errUnexpectedType = errors.New("Metrics does not matches with prometheus")
```
Comment (ober37) [inline utilities.go]:
```suggestion
	var errUnexpectedType = errors.New("Metric name does not comply with Prometheus naming standards")
```

File: utilities.go [hunk 12]
```diff
@@ -111,6 +113,21 @@ func getNodeLabelSet() []string {
...
+	match, _ := regexp.MatchString("[a-zA-Z_:][a-zA-Z0-9_:]*", metricName)
+	if match == true{
```
Comment (ober37) [inline utilities.go]:
```suggestion
	if match == true {
```

File: utilities.go (hunk 10 above)
Comment (ober37) [inline utilities.go]: Only 1 return statement per function, refactor this function

File: exporter.go [hunk 13]
```diff
@@ -198,11 +200,25 @@ func (e *spplugExporter) receiveMessage() func(mqtt.Client, mqtt.Message) {
 
 		// Sparkplug messages contain multiple metrics within them
 		// traverse them and process them
+
 		metricList := pbMsg.GetMetrics()
 
 		for _, metric := range metricList {
-			metricName := metric.GetName()
 
+			metricName, err := getMetricName(metric)
```
Comment (ober37) [inline exporter.go]: Make sure you continue the for loop if you encounter a bad metric, this code will allow the logic to proceed and will try to create the bad metric

File: utilities.go [hunk 14]
```diff
@@ -111,6 +113,21 @@ func getNodeLabelSet() []string {
...
+		return metricName, nil
+		}else {
+			log.Debugf("Error in %v data type format for metric %s\n",errUnexpectedType, metricName)
+			return metricName,errUnexpectedType
```
Comment (ober37) [inline utilities.go]:
```suggestion
			return metricName, errUnexpectedType
```

File: exporter.go (hunk 1 above)
Comment (chaets) [reply exporter.go]: Done

File: exporter.go (hunk 13 above)
Comment (chaets) [reply exporter.go]: Done

File: exporter.go (hunk 2 above)
Comment (chaets) [reply exporter.go]: Done

File: exporter.go (hunk 3 above)
Comment (chaets) [reply exporter.go]: Done

File: utilities.go [hunk 15]
```diff
@@ -113,20 +112,19 @@ func getNodeLabelSet() []string {
...
+		errUnexpectedType = nil
+		} else {
 			log.Debugf("Error in %v data type format for metric %s\n",errUnexpectedType, metricName)
-			return metricName,errUnexpectedType
+			errUnexpectedType = errors.New("Metric name does not comply with Prometheus naming standards")
 		}
-	}
-
-
+	return metricName, errUnexpectedType
```
Comment (ober37) [inline utilities.go]: @chaets In the error situation, I am curious if it makes more sense to set the metricName to nil, this way there is no chance that the caller is getting a malformed name back.

Comment (chaets) [reply utilities.go]: but we won't able to track out what metricName is turning out of prometheus metrics

**Pull request merged**